File: benchmark/fusion/algprog/ex3.14.f
Inductive Tree = tip Int | bin {Tree, Tree};
Inductive PTree = ptip {Int, Int} | pbin {PTree, PTree};

tri = fix (
  \f: Tree -> Compress PTree. \t: Tree.
  match t with
    tip w -> ptip {0, w}
  | bin {l, r} ->
    let step = fix (
      \g: PTree -> PTree. \ys: PTree.
      match ys with
        ptip {n, w} -> ptip {+ n 1, w}
      | pbin {l, r} -> pbin {g l, g r}
      end
    ) in pbin {step (f l), step (f r)}
  end
);

tsum = fix (
  \f: PTree -> Int. \t: PTree.
  match t with
    ptip {n, w} -> * n w
  | pbin {l, r} -> + (f l) (f r)
  end
);

main = \t: Tree. tsum (tri t);

File: ml-benchmark/fusion/algprog/ex3.14.ml
type tree = Tip of int | Bin of tree * tree
type ptree = PTip of int * int | PBin of ptree * ptree

let rec tri t =
  match t with
  | Tip w -> PTip (0, w)
  | Bin (l, r) ->
      let rec step ys =
        match ys with
        | PTip (n, w) -> PTip (n + 1, w)
        | PBin (l2, r2) ->
            let l3 = step l2 in
            let r3 = step r2 in
            PBin (l3, r3)
      in
      let l1 = tri l in
      let r1 = tri r in
      PBin (step l1, step r1)

let rec tsum t =
  match t with
  | PTip (n, w) -> n * w
  | PBin (l, r) ->
      let sl = tsum l in
      let sr = tsum r in
      sl + sr

let program t = tsum (tri t)